packages/node/src/login/oidc/oidcHandlers/ClientCredentialsOidcHandler.ts

```diff
@@ -30,19 +30,102 @@
 import {
   IOidcHandler,
   IOidcOptions,
-  NotImplementedError,
   LoginResult,
+  IStorageUtility,
+  ISessionInfo,
 } from "@inrupt/solid-client-authn-core";
+import { JWK } from "jose";
+import { Issuer } from "openid-client";
+import { inject, injectable } from "tsyringe";
+import { buildAuthenticatedFetch } from "../../../authenticatedFetch/fetchFactory";
+import { configToIssuerMetadata } from "../IssuerConfigFetcher";
+import { getWebidFromTokenPayload } from "../redirectHandler/AuthCodeRedirectHandler";
+import { ITokenRefresher } from "../refresh/TokenRefresher";
 
 /**
  * @hidden
  */
+@injectable()
 export default class ClientCredentialsOidcHandler implements IOidcHandler {
-  async canHandle(_oidcLoginOptions: IOidcOptions): Promise<boolean> {
-    return false;
+  constructor(
+    @inject("node:tokenRefresher") private tokenRefresher: ITokenRefresher,
+    @inject("node:storageUtility") private storageUtility: IStorageUtility
+  ) {}
+
+  async canHandle(oidcLoginOptions: IOidcOptions): Promise<boolean> {
+    return (
+      typeof oidcLoginOptions.client.clientId === "string" &&
+      typeof oidcLoginOptions.client.clientSecret === "string" &&
+      typeof oidcLoginOptions.client.isPublic === "boolean" &&
+      !oidcLoginOptions.client.isPublic
+    );
   }
 
-  async handle(_oidcLoginOptions: IOidcOptions): Promise<LoginResult> {
-    throw new NotImplementedError("ClientCredentialsOidcHandler handle");
+  async handle(oidcLoginOptions: IOidcOptions): Promise<LoginResult> {
+    const issuer = new Issuer(
+      configToIssuerMetadata(oidcLoginOptions.issuerConfiguration)
+    );
+    const client = new issuer.Client({
+      client_id: oidcLoginOptions.client.clientId,
+      client_secret: oidcLoginOptions.client.clientSecret,
+    });
+
+    let dpopKey: JWK.ECKey | undefined;
+
+    if (oidcLoginOptions.dpop) {
+      dpopKey = await JWK.generate("EC", "P-256");
+    }
+
+    const tokens = await client.grant(
+      {
+        grant_type: "client_credentials",
+        client_id: oidcLoginOptions.client.clientId,
+        client_secret: oidcLoginOptions.client.clientSecret,
+        token_endpoint_auth_method: "client_secret_post",
+      },
+      {
+        DPoP: oidcLoginOptions.dpop
+          ? (dpopKey as JWK.ECKey).toJWK(true)
+          : undefined,
+      }
+    );
+
+    if (tokens.access_token === undefined) {
+      throw new Error(
+        `Invalid response from Solid Identity Provider [${
+          oidcLoginOptions.issuer
+        }]: ${JSON.stringify(tokens)} is missing 'access_token'.`
+      );
+    }
+
+    if (tokens.id_token === undefined) {
+      throw new Error(
+        `Invalid response from Solid Identity Provider [${
+          oidcLoginOptions.issuer
+        }]: ${JSON.stringify(tokens)} is missing 'id_token'`
```

Comment:
I'm channelling my inner Pat here:

```suggestion
        }]: [${JSON.stringify(tokens)}] is missing 'access_token'.`
      );
    }

    if (tokens.id_token === undefined) {
      throw new Error(
        `Invalid response from Solid Identity Provider [${
          oidcLoginOptions.issuer
        }]: [${JSON.stringify(tokens)}] is missing 'id_token'`
```

Reply:
Haha I left the square brackets out on purpose, because `JSON.stringify(tokens)` will produce a JSON string enclosed in curly brackets, and I think adding square brackets around that will make it look like an array, which it is not :)

Reply:
Fair enough :)